LFS-752: Create script to package nightly file export

acrowthe opened the pull request:
Every night at 11:30PM, create a file for each subject that had a form created on the current date. Rebased on [LFS-758](https://github.com/ccmbioinfo/lfs/pull/403) as this PR requires date obfuscation.

Testing branch [LFS-752-test](https://github.com/ccmbioinfo/lfs/tree/LFS-752-test) exports every minute instead of nightly.

acrowthe committed:
LFS-752: Create script to package nightly file export
Work in progress.
- Add a script executing at an interval to query for new forms and create a file for each subject with new forms

sdumitriu committed:
LFS-752: Create script to package nightly file export
Task runs correctly

acrowthe committed:
LFS-752: Create script to package nightly file export
Export to file functional

acrowthe committed:
LFS-752: Create script to package nightly file export
Move NightlyExport into cardiac-rehab-resources module

sdumitriu committed:
LFS-752: Create script to package nightly file export
Turn the bundle into a feature to allow the provisioning file to be read

acrowthe committed:
LFS-752: Create script to package nightly file export
Move export to seperate package and set to nightly

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExport.java:
```diff
@@ -0,0 +1,62 @@
+public class NightlyExport
+{
+    /** Default log. */
+    protected static final Logger LOGGER = LoggerFactory.getLogger(NightlyExport.class);
```
```suggestion
    private static final Logger LOGGER = LoggerFactory.getLogger(NightlyExport.class);
```

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExport.java:
```diff
@@ -0,0 +1,62 @@
+    /** Default log. */
+    protected static final Logger LOGGER = LoggerFactory.getLogger(NightlyExport.class);
+
+    /** The Resource Resolver for the current request. */
```
```suggestion
    /** Provides access to resources. */
```

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExport.java:
```diff
@@ -0,0 +1,62 @@
+    @Activate
+    protected void activate(ComponentContext componentContext) throws Exception
+    {
+        LOGGER.error("NightlyExport activating");
```
```suggestion
        LOGGER.info("NightlyExport activating");
```

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExport.java:
```diff
@@ -0,0 +1,62 @@
+        try {
+            this.scheduler.schedule(exportJob, options);
+        } catch (Exception e) {
+            LOGGER.error("NightlyExport Failed to schedule");
```
```suggestion
            LOGGER.error("NightlyExport Failed to schedule: {}", e.getMessage(), e);
```

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,158 @@
+public class NightlyExportTask implements Runnable
+{
+    /** Default log. */
+    protected static final Logger LOGGER = LoggerFactory.getLogger(NightlyExport.class);
```
```suggestion
    private static final Logger LOGGER = LoggerFactory.getLogger(NightlyExportTask.class);
```

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,158 @@
+    /** Default log. */
+    protected static final Logger LOGGER = LoggerFactory.getLogger(NightlyExport.class);
+
+    /** The Resource Resolver for the current request. */
```
```suggestion
    /** Provides access to resources. */
```

k-pace commented:
@acrowthe can we grab user-assigned participant ID ("subject") for the file name rather than UUID?

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,158 @@
+
+            Iterator<Resource> results = resolver.findResources(query, "JCR-SQL2");
+            while (results.hasNext()) {
+                try {
+                    String subjectId = results.next().adaptTo(Node.class).getName();
+                    subjectId = subjectId.substring(subjectId.lastIndexOf("/") + 1);
+                    subjects.add(subjectId);
+                } catch (RepositoryException e) {
+                    LOGGER.warn("Failed to retrieve name of updated subject: {}", e.getMessage(), e);
+                }
```
```suggestion
            while (results.hasNext()) {
                subjects.add(results.next().getPath());
            }
```

acrowthe replied:
Improved: Need the try-catch block to get the subject ideantifier for the file name, but the path is now as suggested.

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,158 @@
+    private SubjectContents getSubjectContents(String subjectId, String requestDateString)
+    {
+        String subjectDataUrl = String.format(
+            "/Subjects/%s.data.deep.bare.-identify.relativeDates.dataFilter:createdAfter=%s",
```
```suggestion
            "%s.data.deep.bare.-identify.relativeDates.dataFilter:createdAfter=%s",
```

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    private final class SubjectContents
```
```suggestion
    private static final class SubjectContents
```

sashaandjic commented:
Lookin good.

acrowthe committed:
LFS-752: Create script to package nightly file export
- Change file name to use participant identifier, not sling name
- Address code review comments

acrowthe committed:
LFS-752: Create script to package nightly file export
- Address access level comment

acrowthe commented:
> @acrowthe can we grab user-assigned participant ID ("subject") for the file name rather than UUID?

Done

sashaandjic commented:
Looking good 
![image](https://user-images.githubusercontent.com/23243244/102284472-35cc5480-3f02-11eb-966b-2f20a87165d3.png)

sashaandjic reviewed (approved):
tested

sashaandjic commented:
@acrowthe External file form seems to be missing.

sashaandjic commented:
@acrowthe I am not seeing External file form in this PR.

sdumitriu committed:
LFS-752: Create script to package nightly file export
No need to turn a resource into a node to read one of its properties

sashaandjic commented:
The only weird thing that I found is that during the day the export started going to Dec 16 folder
![image](https://user-images.githubusercontent.com/23243244/102307193-08979a80-3f32-11eb-88bf-a922dca42dc6.png)

sashaandjic commented:
- Participant status DOB format is wrong ( fixed in LFS-758 )
- External files form is missing ( fixed in LFS-758)
- there is no export if ENV VAR is not set (should be addressed in LFS-758)

k-pace commented:
> The only weird thing that I found is that during the day the export started going to Dec 16 folder
> ![image](https://user-images.githubusercontent.com/23243244/102307193-08979a80-3f32-11eb-88bf-a922dca42dc6.png)

@sashaandjic was it possibly past midnight? You posted this around midnight so wondering if the clock just turned over

sashaandjic commented:
> > The only weird thing that I found is that during the day the export started going to Dec 16 folder
> > ![image](https://user-images.githubusercontent.com/23243244/102307193-08979a80-3f32-11eb-88bf-a922dca42dc6.png)
> 
> @sashaandjic was it possibly past midnight? You posted this around midnight so wondering if the clock just turned over

It actually started happening around 8pm yesterday.

marta- commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,177 @@
+        directory.mkdirs();
+        try (FileWriter file = new FileWriter(path + filename)) {
+            file.write(input.getData());
+            LOGGER.info("Exported {} to {}, size {}B", input.getUrl(), path, directory.length());
```
The `directory.length()` is not useful. We need the size of the downloaded file.

acrowthe committed:
LFS-752: Create script to package nightly file export
- Fix time zone issue causing files to go in the wrong folder
- Fix incorrect file size logging

acrowthe commented:
Resolved timezone issue and rebased on top of the latest LFS-758

sashaandjic commented:
the test branch still has the same problem.
![image](https://user-images.githubusercontent.com/23243244/102422671-46013400-3fd5-11eb-88d7-c1365d67ef58.png)

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,175 @@
+        for (SubjectIdentifier identifier : changedSubjects) {
+            SubjectContents subjectContents = getSubjectContents(identifier.getPath(), requestDateString);
+            if (subjectContents != null) {
+                String filename = String.format("%s_formData_%s.json", identifier.getParticipantId(), fileDateString);
```
Hopefully noone will use `../../` in the subject identifier...

sdumitriu replied:
Yep, this allows writing files anywhere, and may fail if the path leads to a readonly or a non-existent place. The good news is that since it has a suffix, it doesn't allow overwriting things like `/etc/passwd`. Still, I would like to see this path restricted to the output directory by stripping away characters that are not expected in the participant ID. @k-pace @marta- do we know which characters are expected? alphanum is enough, or do we need dashes underscores dots and others?

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,175 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Component
```
I think this isn't needed, since we manually create the instance.

```suggestion
```

k-pace replied:
@sdumitriu they should be alphanumeric only, plan is to use PL1234567 (two letter prefix followed by 7 numbers)

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -0,0 +1,175 @@
+        String path = String.format("%s/cards-exports/%s/", System.getProperty("user.home"), dateString);
+        File directory = new File(path);
+        directory.mkdirs();
+        try {
```
Why did you undo the try-with-resources?

acrowthe replied:
I needed to close the fileWriter to ensure the correct file size could be retrieved and logged. Keeping the try with resource would require an additional state conditional to ensure success isn't logged on a failure, but I could switch to that method if preferred.

acrowthe replied:
Changed over to try with resources + status boolean for logging

acrowthe committed:
LFS-752: Create script to package nightly file export
- Clean patientId before using for filename
- Remove extraneous component annotation
- Refactor filewriter try catch

acrowthe commented:
Patient identifiers are now stripped of all non-alphanumeric characters before being used in the filename (`[^A-Za-z0-9]`)

sdumitriu commented on cardiac-rehab-resources/backend/src/main/java/ca/sickkids/ccm/lfs/cardiacrehab/internal/export/NightlyExportTask.java:
```diff
@@ -163,13 +169,19 @@ private void output(SubjectContents input, String filename, String dateString)
         }
+        if (success) {
+            LOGGER.info(
+                "Exported {} to {}, size {}B",
```
Do we really need the size? The code could be simplified if we don't print it, especially since its in an ignored log message.

marta- replied:
This was for auditing purposes. It is not crucial if it's causing that much trouble.

acrowthe replied:
Byte size logging removed

sashaandjic reviewed (approved):
tested

acrowthe commented:
Date folders performed as expected across the UTC date boundary on a full ubuntu VM: I suspect the issues @sashaandjic has been encountering with exports going into the wrong folder is due to ubuntu for WSL.

acrowthe committed:
LFS-798: Optimize export code to avoid unnecessary re-exporting
- Fix subject with multiple forms being duplicated in the subjects set
  by implementing equals() and hashCode() methods

acrowthe commented:
Did notice a bug where a subject with multiple forms was being processed multiple times. Not critical since no data is being missed and the duplicate output files overwrite each other, but will cause the exports to take longer than necessary.

sashaandjic commented:
> Did notice a bug where a subject with multiple forms was being processed multiple times. Not critical since no data is being missed and the duplicate output files overwrite each other, but will cause the exports to take longer than necessary.

Created JIRA task to address this issue: [LFS-798](https://phenotips.atlassian.net/browse/LFS-798)

acrowthe committed:
LFS-752: Create script to package nightly file export
- Clean up logging

acrowthe commented:
> Created JIRA task to address this issue: [LFS-798](https://phenotips.atlassian.net/browse/LFS-798)

This issue has now been resolved. Subjects are only exported once, even with multiple forms

sashaandjic reviewed (approved):
re-tested
